waspc/src/Analyzer/Parser/Util.hs

```diff
@@ -0,0 +1,53 @@
+module Analyzer.Parser.Util
+  ( ParseState (..),
+    initialState,
+    Parser,
+    updatePosn,
+    putInput,
+    AlexInput,
+  )
+where
+
+import Analyzer.Parser.Syntax (ParseError, Posn (..))
+import Control.Monad.Trans.Except (Except)
+import Control.Monad.Trans.State.Lazy (StateT, get, modify)
+import Data.Word (Word8)
+
+-- | Tracks state of the parser, which is
+--   - Current line/column position
+--   - Current input to the lexer
+data ParseState = ParseState
+  { psPosn :: Posn,
+    psInput :: AlexInput
```

Comment:
s/psInput/parserRemainingInput

And then it is somewhat weird to see `AlexInput` here.